File: cypress/integration/properties.js
before(() => {
  cy.visit('/')
})

const path = require("path");

describe('Properties Pane', () => {

  it('check download button', () => {
      cy.run('properties_basic')
          .get('.layout .react-grid-item').first()
          .find('button[title="save"]').click()
      const downloadsFolder = Cypress.config("downloadsFolder");
      cy.readFile(path.join(downloadsFolder, "visdom_properties.json")).should("exist");
  });


  it('properties_callbacks', () => {
      cy.run('properties_callbacks', {asyncrun: true})

      cy.get('input[value="initial"]').first().clear().type("changed{enter}")
      cy.get('.layout .react-grid-item .content-text').first().contains("Updated: Text input => changed")
      cy.get('input[value="changed_updated"]')

      cy.get('input[value="12"]').first().clear().type("42{enter}")
      cy.get('.layout .react-grid-item .content-text').first().contains("Updated: Number input => 42")
      cy.get('input[value="420"]')

      cy.get('button').contains("Start").click()
      cy.get('.layout .react-grid-item .content-text').first().contains("Updated: Button => clicked")

      cy.get('input[type="checkbox"]').first().should("be.checked").click()
      cy.get('.layout .react-grid-item .content-text').first().contains("Updated: Checkbox => False")
      cy.get('input[type="checkbox"]').first().should("not.be.checked")

      cy.get('select').first().should("have.value", "1").select('Red')
      cy.get('.layout .react-grid-item .content-text').first().contains("Updated: Select => 0")
      cy.get('select').first().should("have.value", "0").select('Blue')
      cy.get('.layout .react-grid-item .content-text').first().contains("Updated: Select => 2")
      cy.get('select').first().should("have.value", "2")

  })

})

File: example/components/properties.py
from components.text import text_callbacks

# Properties window
def properties_basic(viz, env, args):
    properties = [
        {'type': 'text', 'name': 'Text input', 'value': 'initial'},
        {'type': 'number', 'name': 'Number input', 'value': '12'},
        {'type': 'button', 'name': 'Button', 'value': 'Start'},
        {'type': 'checkbox', 'name': 'Checkbox', 'value': True},
        {'type': 'select', 'name': 'Select', 'value': 1, 'values': ['Red', 'Green', 'Blue']},
    ]
    properties_window = viz.properties(properties, env=env)
    return properties, properties_window


def properties_callbacks(viz, env, args):
    callback_text_window = text_callbacks(viz, env, args)
    properties, properties_window = properties_basic(viz, env, args)

    def properties_callback(event):
        if event['event_type'] == 'PropertyUpdate':
            prop_id = event['propertyId']
            value = event['value']
            if prop_id == 0:
                new_value = value + '_updated'
            elif prop_id == 1:
                new_value = value + '0'
            elif prop_id == 2:
                new_value = 'Stop' if properties[prop_id]['value'] == 'Start' else 'Start'
            else:
                new_value = value
            properties[prop_id]['value'] = new_value
            viz.properties(properties, win=properties_window, env=env)
            viz.text("Updated: {} => {}".format(properties[event['propertyId']]['name'], str(event['value'])),
                     win=callback_text_window, append=True, env=env)

    viz.register_event_handler(properties_callback, properties_window)
